src/FunderMaps.Core.Authentication/SignInService.cs

```diff
@@ -0,0 +1,226 @@
+        {
+            var passwordHash = await UserRepository.GetPasswordHashAsync(id);
+            return PasswordHasher.IsPasswordValid(passwordHash, password);
+        }
+
+        /// <summary>
+        ///     Set the password for the user.
+        /// </summary>
+        /// <param name="id">The user id whose password should be set.</param>
+        /// <param name="password">The password to be set on the user.</param>
+        public virtual async Task SetPasswordAsync(Guid id, string password)
+        {
+            var passwordHash = PasswordHasher.HashPassword(password);
+            await UserRepository.SetPasswordHashAsync(id, passwordHash);
+        }
+
+        /// <summary>
+        ///     Attempts to sign in the specified <paramref name="principal"/>.
+        /// </summary>
+        /// <param name="principal">The principal to sign in.</param>
+        /// <returns>Instance of <see cref="SignInContext"/>.</returns>
+        public virtual async Task<SignInContext> SignInAsync(ClaimsPrincipal principal, string authenticationType)
+        {
+            if (principal == null)
+            {
+                throw new ArgumentNullException(nameof(principal));
+            }
+
+            var (user, tenant) = PrincipalProvider.GetTenantUser<User, Organization>(principal);
+            if (user == null || tenant == null)
+            {
+                return SignInContext.Failed;
+            }
+
+            var claim = principal.FindFirst(FunderMapsAuthenticationClaimTypes.TenantRole);
+            if (claim == null)
+            {
+                return SignInContext.Failed;
+            }
+
+            return await SignInAsync(user, tenant, Enum.Parse<OrganizationRole>(claim.Value), authenticationType);
+        }
+
+        /// <summary>
+        ///     Attempts to sign in the specified <paramref name="user"/> and <paramref name="password"/> combination.
+        /// </summary>
+        /// <param name="user">The user to sign in.</param>
+        /// <returns>Instance of <see cref="SignInContext"/>.</returns>
+        public virtual async Task<SignInContext> SignInAsync(IUser user, ITenant tenant, OrganizationRole organizationRole, string authenticationType)
+        {
+            if (user == null)
+            {
+                throw new ArgumentNullException(nameof(user));
+            }
+
+            if (tenant == null)
+            {
+                throw new ArgumentNullException(nameof(tenant));
+            }
+
+            var result = await CanSignInAsync(user.Id);
+            if (result != SignInContext.Success)
+            {
+                return result;
+            }
+
+            Logger.LogInformation(1, $"User {user} sign in was successful.");
+
+            return new SignInContext(
+                result: AuthResult.Success,
+                principal: PrincipalProvider.CreateTenantUserPrincipal(user, tenant, organizationRole, authenticationType));
```

Comment:
This line can throw an exception but the logger has already logged a success. I suggest first getting all values, then logging success, then returning.